src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseHeaderReader.cs

```diff
@@ -0,0 +1,168 @@
+                }
+
+                return true;
+            }
+
+            name = null;
+            value = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Reads lines separated by "\r\n".
+        /// </summary>
+        /// <returns>true if the next line was read successfully, or false if all characters have been read.</returns>
+        public bool ReadLine()
+        {
+            int startIndex;
+            int length;
+            return ReadLine(out startIndex, out length);
+        }
+
+        /// <summary>
+        /// Reads lines separated by "\r\n".
+        /// </summary>
+        /// <param name="startIndex">The start of the line segment.</param>
+        /// <param name="length">The length of the line segment.</param>
+        /// <returns>true if the next line was read successfully, or false if all characters have been read.</returns>
+        private bool ReadLine(out int startIndex, out int length)
+        {
+            Debug.Assert(_buffer != null);
+
+            int i = _position;
+
+            while (i < _length)
+            {
+                char ch = _buffer[i];
+                if (ch == '\r')
+                {
+                    int next = i + 1;
+                    if (next < _length && _buffer[next] == '\n')
+                    {
+                        startIndex = _position;
+                        length = i - _position;
+                        _position = i + 2;
+                        return true;
+                    }
+                }
+                i++;
+            }
+
+            if (i > _position)
+            {
+                startIndex = _position;
+                length = i - _position;
+                _position = i;
+                return true;
+            }
+
+            startIndex = 0;
+            length = 0;
+            return false;
+        }
+
+        private static bool TryGetKnownHeaderValue(string name, char[] array, int startIndex, int length, out string value)
+        {
+            Debug.Assert(name != null);
+            CharArrayHelpers.DebugAssertArrayInputs(array, startIndex, length);
+
+            // Do a really quick reference equals check to see if name is the same object as
+            // HttpKnownHeaderNames.ContentEncoding, in which case the value is very likely to
+            // be either "gzip" or "deflate".
+            if (object.ReferenceEquals(name, HttpKnownHeaderNames.ContentEncoding))
+            {
+                if (CharArrayHelpers.EqualsOrdinalAsciiIgnoreCase(Gzip, array, startIndex, length))
+                {
+                    value = Gzip;
+                    return true;
+                }
+                else if (CharArrayHelpers.EqualsOrdinalAsciiIgnoreCase(Deflate, array, startIndex, length))
+                {
+                    value = Deflate;
+                    return true;
+                }
+            }
+
+            value = null;
+            return false;
```

Comment:
Does this method need to be a Try* method, or could it just return the string, either a cached one or an allocated one?